### DISCUSSION TITLE
How to put nulls in the y-axis series?

### DISCUSSION THREAD

**Discussion by vivekgupta4Git** [patrykandpatrick/vico#726] [Questions] [answered]
I am creating a Line Chart and the x series is in the range of 0..11 (representing months). The Y's series may contain only one value.   For this, I have created an arrayOfNulls for y axis series. For example, Let say we receive the data from api in the format :
`
{"graphData":[{"data":0.3,"date":"2024-05"},{"data":0.58,"date":"2024-04"}],"totalAvg":0.53}
`
<img width="363" alt="Screenshot 2024-05-24 at 4 48 21 PM" src="https://github.com/patrykandpatrick/vico/assets/91813403/84f74209-bd31-4ec5-beb5-f1db7eb78c1b">

So My x-axis is 1..12  and at 5th position of arrayOfNulls 0.3 will get set and so on. But the series function doesn't accept this array !
So I tried taking floatArray of size 12 ( with default value 0.0f ) and I set the 0.58f at 4th position , 0.3 at 5th position of this array. Series function accepts this but I don't want to draw points for 0.0f. 
I tried putting Float.NaN in the rest of floatArray elements but then there is no values drawn for y series. 
<img width="363" alt="Screenshot 2024-05-24 at 4 47 00 PM" src="https://github.com/patrykandpatrick/vico/assets/91813403/06080404-c740-4d65-9782-208d18b905d1">

It should draw a point on the graph for that particular value  and if there is more than one value it should draw a line. How to achieve this?

code for point connector :

> val pointConnector = object : LineCartesianLayer.LineSpec.PointConnector {
        override fun connect(path: android.graphics.Path, prevX: Float, prevY: Float, x: Float, y: Float, horizontalDimensions: HorizontalDimensions, bounds: RectF) {
            with(path) {
                lineTo(prevX, prevY)
               }
            }
        }

**Comment by vivekgupta4Git** [comment]
This was essential for the app. it's sad that i had to replace this awesome library with mpChart.

**Comment by patrickmichalik** [comment, marked as answer]
Hello, @vivekgupta4Git! You can do this as follows:

1. Populate the series with only those values for which points should be shown:

    ```kt
    series(x = listOf(4, 5), y = listOf(0.58, 0.3))
    ```
2. Set the _x_ range via `AxisValueOverrider`:

    ```kt
    AxisValueOverrider.fixed(minX = 0f, maxX = 11f)
    ```

The result is shown below. This works for any number of points, including one.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/3f2afdbd-20c4-418d-923c-de0ce982ee3f" width="720" />

The custom `LineCartesianLayer.LineSpec.PointConnector` should use `lineTo(x, y)`, not `lineTo(prevX, prevY)`. Alternatively, for sharp corners, the following can be used:

```kt
DefaultPointConnector(cubicStrength = 0f)
```

**Reply by vivekgupta4Git** [reply]
It works. Thanks ! but I have one more question. How can I show all the labels of x- axis? I have requirement to show day-wise data as well. So x series starts from 1 to 31 but the graph shows 1..15.29..  instead of showing all labels.  
<img width="354" alt="Screenshot 2024-05-27 at 10 49 12 PM" src="https://github.com/patrykandpatrick/vico/assets/91813403/acf37c35-f38c-4116-a296-3df1bb3f4650">

**Reply by patrickmichalik** [reply]
You’re welcome, @vivekgupta4Git! You can do this by overriding the _x_ step accordingly:

```kt
getXStep = { 1f }
```

`CartesianChartHost` has a `getXStep` parameter, and `CartesianChartView` has a `getXStep` property. I was planning on mentioning locking the _x_ step initially—it’s a good idea for the first chart too—but it slipped my mind. Sorry about that.

Incidentally, what’s seen in the screenshot seems to point to an oversight on our end. The default _x_ step should be such that each series _x_ value can be expressed as a sum of the _x_ range’s lower bound and a multiple of the _x_ step. It looks like range overrides aren’t correctly accounted for, resulting in suboptimal default behavior. We’ll look into it. Regardless, locking the _x_ step is appropriate here, and it should produce the desired result.
